packages/camera/camera/example/ios/RunnerTests/FLTCamPhotoCaptureTests.m

```diff
@@ -0,0 +1,172 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@import camera;
+@import camera.Test;
+@import AVFoundation;
+@import XCTest;
+#import <OCMock/OCMock.h>
+
+@interface FLTCamPhotoCaptureTests : XCTestCase
+
+@end
+
+@implementation FLTCamPhotoCaptureTests
+
+- (void)testCaptureToFile_savePhotoDelegateReferencesMustBeAccessedOnCaptureSessionQueue {
+  XCTestExpectation *setReferenceExpectation =
+      [self expectationWithDescription:
+                @"FLTSavePhotoDelegate references must be set on capture session queue."];
+  XCTestExpectation *clearReferenceExpectation =
+      [self expectationWithDescription:
+                @"FLTSavePhotoDelegate references must be cleared on capture session queue."];
+
+  dispatch_queue_t captureSessionQueue = dispatch_queue_create("capture_session_queue", NULL);
+  const char *captureSessionQueueSpecific = "capture_session_queue";
+  dispatch_queue_set_specific(captureSessionQueue, captureSessionQueueSpecific,
+                              (void *)captureSessionQueueSpecific, NULL);
+  FLTCam *cam = [self createFLTCamWithCaptureSessionQueue:captureSessionQueue];
+
+  // settings.uniqueID is used as the key for `inProgressSavePhotoDelegates` dictionary
+  AVCapturePhotoSettings *settings = [AVCapturePhotoSettings photoSettings];
+  id mockSettings = OCMClassMock([AVCapturePhotoSettings class]);
+  OCMStub([mockSettings photoSettings]).andReturn(settings);
+
+  // We need to make sure the delegate reference is actually saved in a real dictionary, so that we
+  // can call its completion handler later. Must use a partial mock, in order to forward invocation
+  // to the real object.
+  id mockDelegates = OCMPartialMock([NSMutableDictionary dictionary]);
+  OCMStub([mockDelegates setObject:OCMOCK_ANY forKeyedSubscript:OCMOCK_ANY])
+      .andDo(^(NSInvocation *invocation) {
+        if (dispatch_get_specific(captureSessionQueueSpecific)) {
```

Comment:
Why is all of this mocking necessary; couldn't you do essentially this same logic in the code itself with `NSAssert`s (instead of comments saying what the current queue is), and then have the unit test just exercise the actual codepath?

Reply:
Wow thank you for this explanation. It was very helpful! 

I will delete some of those "queue-verifying" unit tests for the ones that did not cause externally visible bugs. As long as they have other tests that goes through `NSAssert`, we should be confident.